File: typescript/tests/accessor-inventory.test.ts
/**
 * Accessor-roster inventory: every generated service must be *reachable*.
 *
 * `check-typescript-service-drift.sh` proves the generated service directory
 * matches the spec, and `make check-service-inventory-parity` proves the five
 * generators emitted the same service set. Neither asks whether the
 * hand-written wiring exposes what was generated — and TypeScript has four
 * hand-maintained renderings of one roster:
 *
 *   1. the `import` block in src/client.ts
 *   2. the `BasecampClient` interface properties
 *   3. the `defineService` calls
 *   4. the export blocks in src/index.ts
 *
 * A service missing from any of the first three is unreachable at runtime; a
 * service missing from the fourth is unreachable to a *consumer*, because
 * package.json exports only the package root and `./oauth`. Nothing fails in
 * either case: nothing references the missing accessor, so compilation is
 * clean. Python shipped `gauges` unreachable for about a year that way (#732,
 * #755). This is that question, asked of both surfaces.
 *
 * Two properties of the derivation are deliberate:
 *
 *   - The roster comes from the generated directory, never a literal. A literal
 *     would have to be edited by the same person who forgot the accessor.
 *   - Resolution goes through the modules, not their source text. Two
 *     `defineService` calls span multiple lines and one index.ts export is a
 *     single line, so a line-oriented regex over either file silently misses
 *     them.
 */

import { describe, expect, it } from "vitest";
import fs from "node:fs";
import { fileURLToPath } from "node:url";

import { createBasecampClient } from "../src/client.js";
import * as generatedServices from "../src/generated/services/index.js";
import * as rootExports from "../src/index.js";
import { BaseService } from "../src/services/base.js";

const GENERATED_SERVICES_DIR = fileURLToPath(new URL("../src/generated/services", import.meta.url));

/**
 * Non-vacuity floor. Every assertion below is `it.each` over a derived roster,
 * so an import or path change that yielded an empty roster would make all of
 * them vacuously true. The floor is well under the real count (53) so it does
 * not become a second constant to maintain.
 */
const MIN_GENERATED_SERVICES = 40;

/** Services on the flat client that are not generated from the spec. */
const HAND_WRITTEN_ACCESSORS = new Set(["authorization"]);

/** `*Service` value exports of index.ts that are not generated services. */
const NON_SERVICE_EXPORTS = new Set(["BaseService", "AuthorizationService"]);

const toCamelCase = (kebab: string): string => kebab.replace(/-([a-z])/g, (_, c: string) => c.toUpperCase());

const toPascalCase = (kebab: string): string => {
  const camel = toCamelCase(kebab);
  return camel.charAt(0).toUpperCase() + camel.slice(1);
};

/** Every generated service module, by kebab-case basename. */
const GENERATED_MODULES: string[] = fs
  .readdirSync(GENERATED_SERVICES_DIR)
  .filter((f) => f.endsWith(".ts") && f !== "index.ts")
  .map((f) => f.slice(0, -".ts".length))
  .sort();

/** module basename -> { accessor on the client, exported class name }. */
const ROSTER = GENERATED_MODULES.map((module) => ({
  module,
  accessor: toCamelCase(module),
  className: `${toPascalCase(module)}Service`,
}));

const serviceClass = (className: string): Function => {
  const cls = (generatedServices as unknown as Record<string, unknown>)[className];
  // Asserted rather than assumed: the accessor and class names are derived from
  // the filename by one rule each, and this is where an irregular spelling the
  // generator might introduce surfaces as a named failure instead of an
  // `undefined` two assertions later.
  expect(typeof cls, `generated/services/index.ts does not export ${className}`).toBe("function");
  return cls as Function;
};

const newClient = () => createBasecampClient({ accountId: "12345", accessToken: "test-token" });

const propertyOf = (target: object, name: string): unknown => (target as Record<string, unknown>)[name];

describe("generated service reachability", () => {
  it("derives a non-empty roster from the generated directory", () => {
    expect(GENERATED_MODULES.length).toBeGreaterThan(MIN_GENERATED_SERVICES);
    // The generated barrel is emitted by the same loop that writes the modules,
    // so a disagreement here means the directory holds something the generator
    // did not produce.
    expect(Object.keys(generatedServices).length).toBe(GENERATED_MODULES.length);
  });

  describe("client.ts wiring", () => {
    it.each(ROSTER)("client.$accessor is a $className", ({ module, accessor, className }) => {
      const client = newClient();
      const cls = serviceClass(className);

      const service = propertyOf(client, accessor);
      expect(service, `BasecampClient has no \`${accessor}\`; generated/services/${module}.ts defines that service but client.ts does not wire it`).toBeDefined();
      // Presence alone would pass if `gauges` returned the wrong service, so
      // resolve it and check the class it actually yields. The six hand-written
      // composites (§18) subclass their generated service, so `instanceof`
      // holds for them too.
      expect(service).toBeInstanceOf(cls);
    });

    it("has no accessor without a generated service", () => {
      const client = newClient();
      // Every service accessor is a BaseService instance, which is what
      // separates them from the raw openapi-fetch verbs and the client's own
      // properties without naming those individually.
      const wired = Object.keys(client)
        .filter((key) => propertyOf(client, key) instanceof BaseService)
        .filter((key) => !HAND_WRITTEN_ACCESSORS.has(key))
        .sort();

      expect(wired).toEqual(ROSTER.map((r) => r.accessor).sort());
    });
  });

  describe("index.ts exports", () => {
    // Needs its own assertion: a missing export block is invisible to the
    // runtime checks above, which reach the service through client.ts. An
    // in-repo importer never notices, and only a consumer of the published
    // package does.
    it.each(ROSTER)("index.ts exports $className", ({ className }) => {
      const cls = serviceClass(className);
      const exported = propertyOf(rootExports, className);

      expect(exported, `src/index.ts does not export ${className}; consumers cannot import it through any supported path`).toBeDefined();
      // Six index.ts blocks export types only, their class coming from a
      // hand-written extension subclass exported separately — so the export is
      // required to *be* the generated class or to extend it, not to be
      // identical to it.
      const isTheServiceClass = exported === cls || (exported as { prototype?: unknown })?.prototype instanceof cls;
      expect(isTheServiceClass, `src/index.ts exports ${className}, but it is neither the generated class nor a subclass of it`).toBe(true);
    });

    it("exports no service class without a generated service", () => {
      const exportedServices = Object.keys(rootExports)
        .filter((name) => name.endsWith("Service") && !NON_SERVICE_EXPORTS.has(name))
        .sort();

      expect(exportedServices).toEqual(ROSTER.map((r) => r.className).sort());
    });
  });
});

File: ruby/test/basecamp/accessor_inventory_test.rb
# frozen_string_literal: true

require "test_helper"

# Accessor-roster inventory: every generated service must be *reachable*.
#
# check-ruby-service-drift.sh proves the generated service directory matches the
# spec, and `make check-service-inventory-parity` proves the five generators
# emitted the same service set. Neither asks whether AccountClient exposes what
# was generated — the accessors in client.rb's `# @!group Services` block are
# hand-written, and a missing one breaks nothing: nothing references it, so the
# suite stays green and the service is simply unreachable. Python shipped
# `gauges` that way for about a year (#732, #755).
#
# The roster is derived from the generated directory rather than typed out here,
# so a newly generated service enters this test the moment it is generated and
# fails until someone wires an accessor. A hand-copied literal would have to be
# edited by the same person who forgot the accessor.
#
# Ruby needs no subclass allowance, unlike Python and TypeScript: the five
# hand-written composites are `prepend`ed onto the generated classes by the
# zeitwerk `on_load` hooks in lib/basecamp.rb, so the generated constant is
# still the accessor's exact class.
class AccessorInventoryTest < Minitest::Test
  include TestHelper

  GENERATED_SERVICES_DIR = File.expand_path("../../lib/basecamp/generated/services", __dir__)

  # Hand-written infrastructure living beside the generated services.
  NON_GENERATED_SERVICE_FILES = [ "base_service.rb" ].freeze

  # Zero-arity public methods on AccountClient that are not service accessors.
  INFRASTRUCTURE_METHODS = %i[account_id config http hooks].freeze

  # Non-vacuity floor. Every assertion below iterates a derived roster, so a
  # path change that yielded an empty roster would make all of them vacuously
  # true. Kept well under the real count (53) so it is not a second constant to
  # maintain.
  MIN_GENERATED_SERVICES = 40

  # accessor name (snake_case) => generated class constant name.
  ROSTER = Dir.children(GENERATED_SERVICES_DIR)
    .select { |f| f.end_with?("_service.rb") }
    .reject { |f| NON_GENERATED_SERVICE_FILES.include?(f) }
    .to_h do |file|
      accessor = File.basename(file, "_service.rb")
      [ accessor.to_sym, "#{accessor.split("_").map(&:capitalize).join}Service" ]
    end
    .freeze

  def test_roster_is_derived_and_not_empty
    assert_operator ROSTER.size, :>, MIN_GENERATED_SERVICES,
      "roster derivation yielded #{ROSTER.size} services; every assertion in this file would be vacuous"
  end

  def test_every_generated_service_has_an_accessor
    account = create_client.for_account(account_id)

    ROSTER.each_key do |accessor|
      assert_respond_to account, accessor,
        "AccountClient has no `#{accessor}` accessor; generated/services/#{accessor}_service.rb " \
        "defines that service but client.rb does not wire it"
    end
  end

  def test_each_accessor_resolves_to_that_service
    account = create_client.for_account(account_id)

    ROSTER.each do |accessor, class_name|
      klass = Basecamp::Services.const_get(class_name)

      # Presence alone would pass if `gauges` returned the wrong service, so
      # resolve it and check the class it actually yields.
      assert_instance_of klass, account.public_send(accessor),
        "AccountClient##{accessor} does not return #{class_name}"
    end
  end

  def test_generated_classes_are_all_constants
    # Guards the class-name derivation itself: one rule turns a filename into a
    # constant name, and an irregular spelling should fail here by name rather
    # than as a NameError buried in the identity test above.
    ROSTER.each do |accessor, class_name|
      assert Basecamp::Services.const_defined?(class_name),
        "#{accessor}_service.rb does not define Basecamp::Services::#{class_name}"
    end
  end

  def test_no_accessor_without_a_generated_service
    # The other direction: an accessor left behind by a removed or renamed
    # service. Every service accessor is a zero-arity public method defined
    # directly on AccountClient; the four infrastructure readers are the only
    # others, and they are named rather than derived so a new one has to be
    # acknowledged here.
    #
    # The class is read off an instance rather than named as a constant.
    # `Basecamp::AccountClient` is declared inside client.rb alongside
    # `Basecamp::Client`, so zeitwerk has no autoload registered for it: naming
    # it directly resolves only once something else has already loaded that
    # file, which under minitest's random seed is a coin flip (it raised
    # NameError on the first seed that ran this test first).
    account_client_class = create_client.for_account(account_id).class

    zero_arity = account_client_class.public_instance_methods(false)
      .select { |m| account_client_class.instance_method(m).arity.zero? }

    assert_equal ROSTER.keys.sort, (zero_arity - INFRASTRUCTURE_METHODS).sort
  end

  def test_authorization_is_not_an_account_scoped_service
    # Stated so the reverse assertion above cannot be quietly loosened to
    # accommodate it: `authorization` is an OAuth service on the outer Client,
    # outside the account-scoped roster entirely. Asserted against the same
    # method list that assertion enumerates, not against `respond_to?`, so the
    # two cannot disagree about where the accessor lives.
    client = create_client

    assert_respond_to client, :authorization
    assert_not_includes client.for_account(account_id).class.public_instance_methods(false), :authorization
  end
end
